refactor: use python classes

olevski opened the pull request:
@ableuler let me know what you think.

The biggest difference from what we discussed is that there is essentially one big class that represents a JH server. This made sense to me compared to trying to make specific classes for every endpoint.

Also it made sense for the authenticated user object returned by the `authenticated` decorator to be an input to the server class. So because of that I could not make the class be the direct result of the schema. But it is very close, what we have in most cases now is schema -> dictionary -> server class. Then the server class has methods to start a server, delete it, get logs, get a summary of the k8s pod (i.e. used in /servers and /servers/name).

Let me know what you think.

I also adjusted the tests to have fixtures that try to mimic k8s. So as a test runs there is a fixture that keeps track of which pods were created by JH. I.e. like a very dumb mock version of k8s.

I tested the acceptance tests with docker and they pass but it would not be a bad idea to deploy this PR.

/deploy

olevski commented:
/deploy

ableuler commented:
> The biggest difference from what we discussed is that there is essentially one big class that represents a JH server. This made sense to me compared to trying to make specific classes for every endpoint.

That is definitely the better solution. In fact there might have been a misunderstanding. When suggesting that there should be a class-to-schema correspondence, I was thinking of consolidating the schemas too rather than having a different server class for each HTTP method.

I think the most "unorthodox" aspect of the current implementation is that we have a server class and we have some server schemas, but we don't really use the schemas for (de-)serialization server objects. I'm not yet 100% sure how exactly this would work, but by defining a [post_load method](https://marshmallow.readthedocs.io/en/stable/quickstart.html#deserializing-to-objects) on the schemas it should be possible do de-serialize dictionaries directly into server objects, thereby also validating the input dictionary against the schema. Similarly, it would be nice if the response of the POST request when creating a new session was simply the marshmallow dump of the created server object.

olevski commented:
@ableuler I pushed new code that should clear all conflicts and lingering issues. I am troubleshooting one of the acceptance tests which seems to fail because the "Templates" dropdown on the new project page takes a while to populate. This issue is unrelated to the code changes here.

```
Screenshot saved to /tests/target/20210224154154-DatasetModificationFlowSpec.png
[info] DatasetModificationFlowSpec:
[info] Scenario: From the UI the user can modify a dataset and only interact with its latest version *** FAILED ***
[info]   The code passed to eventually never returned normally. Attempted 1 times over 1.23015771 minutes. Last failure message: The code passed to eventually never returned normally. Attempted 349 times over 1.000906195 minutes. Last failure message: Template option not found.. (NewProjectPage.scala:44)
[info]   Given user is not logged in 
[info]   When user clicks on the Login button 
[info]   Then they should get into the Login Page 
[info]   When user clicks on the provider login page 
[info]   And enters credentials and logs in 
[info]   Then they should get into the Welcome page 
[info]   Given a new renku project 
[info]   When user clicks on the 'New Project' menu item 
[info]   Then the New Project page gets displayed 
[info]   When user fills in and submits the new project details form 
```

ableuler reviewed (changes requested):
This is a big step in the right direction! Also, the use of the `use_kwargs` decorator to validate the request data instead of doing this through `schema.load()`seems perfectly fine to me now that I understand the new implementation a bit better. At some point we should be able to use the counterpart `marshal_with` decorator for the response serialization.

I see two major improvements to the proposed refactoring:

1. It would good to use the server class in the handling of all server-related API requests. Specifically, this would mean that for a GET requests to a `/servers`, we would query K8s just as you would query a DB in a kind of "textbook" REST API, from the response we create instances of the server class (basically the equivalent to the object-relational mapping in the database case) which are then serialized and returned to the client. 

2. I think the server class is a bit overloaded and would greatly benefit from an additional class that models the user and handles all the user-related auth information. I would probably also move the methods to create a server for a user and query server(s) for a user to the user class. The server class on the other hand should probably have all the logic for managing the pod that represents a server object in the k8s cluster.

olevski commented on renku_notebooks/api/classes/server.py:
```diff
@@ -0,0 +1,416 @@
+            )
+        return logs
+
+    def k8s_summary(self):
```
@ableuler I was able to move this serailizer into the pre_dump method of the server schema. So the code is not repeated. The pre_dump method expects a server object, it takes the pod spec and converts it into the format the UI expects.

olevski commented on renku_notebooks/api/classes/server.py:
```diff
@@ -0,0 +1,416 @@
+        self._auth = HubOAuth(
+            api_token=os.environ.get("JUPYTERHUB_API_TOKEN"), cache_max_age=60
+        )
+        self._user = user
```
@ableuler I added a user class. I think it worked out really well. A lot of random functions used for authentication naturally were absorbed by this class.

olevski commented on renku_notebooks/api/classes/server.py:
```diff
@@ -0,0 +1,416 @@
+            )
+
+    @classmethod
+    def from_server_name(cls, user, server_name):
```
@ableuler the server class now has two constructors - `from_server_name` and `from_pod`. They work as you suggested.

olevski commented:
@ableuler I made the changes you suggested. I think things worked out really well. Also as you mentioned we do not need the custom response validation decorator anymore. I am having some issues with the acceptance tests (when running them locally). However I cannot replicate the failures that the acceptance tests report when I try to do the same thing. I will ask for help on this if I cannot figure it out. Let me know what you think.

rokroskar commented:
@olevski the acceptance tests running in CI passed...

rokroskar commented on renku_notebooks/api/classes/server.py:
```diff
@@ -0,0 +1,399 @@
+)
+
+
+class Server:
```
Just a naming suggestion, feel free to ignore... but casually glancing at this code I expected this to represent the jupyterhub server, not a user session. Maybe it could be called a `JupyterSession` or `UserSession` or `UserServer` or something similar?

olevski replied:
I see what you mean, I will update to UserServer.

olevski commented:
> @olevski the acceptance tests running in CI passed...

@rokroskar I saw that. I think the issues I am having locally are not really a problem. As long as these tests are passing I will not waste time trying to troubleshoot the local tests that I run through docker.

ableuler reviewed (changes requested):
I really like how this is coming together. For me the big picture of the main classes, schemas, request and response validation and also serializers (to the extent we can streamline them without breaking the API) is there. I still have some suggestions for changes, but those are at a much more detailed level.

olevski commented on renku_notebooks/api/classes/user.py:
```diff
@@ -0,0 +1,78 @@
+        return {self.auth.auth_header_name: f"token {self.auth.api_token}"}
+
+    @property
+    def user(self):
```
I removed this

olevski commented on renku_notebooks/api/classes/user.py:
```diff
@@ -0,0 +1,78 @@
+        response = requests.get(
+            f"{self.prefix}/users/{self.user['name']}", headers=self.headers
+        )
+        return json.loads(response.text)
```
I made this change.

olevski commented on renku_notebooks/api/classes/server.py:
```diff
@@ -0,0 +1,399 @@
+        self._jupyterhub_origin = os.environ.get("JUPYTERHUB_ORIGIN", "")
+
+    @staticmethod
+    def make_server_name(namespace, project, branch, commit_sha):
```
@ableuler I removed `renku_notebooks/util/jupyterhub_.py` that was not used anywhere at all. I just completely missed it.

As for making this a property I would like to keep it as a static method because it allows me to use it in the tests here:
https://github.com/SwissDataScienceCenter/renku-notebooks/pull/541/files#diff-205841df8f264252b295f3101d045d55f31a0044526da199256b89e12d9518dfR29

And even though the tests that use `make_servername` so far are not that fancy - we can write more in the future and using this would be useful. Also it truly does not depend on anything else I just thought it made sense to be part of the server class. Lastly because we use this as the name of the JH server then JH enforces this result to be unique. So by keeping it a static method we can easily see what we and JH considers to be a unique server.

olevski commented on renku_notebooks/api/classes/server.py:
```diff
@@ -0,0 +1,399 @@
+            self.namespace, self.project, self.branch, self.commit_sha
+        )
+
+    def _get_environment_vars(self):
```
@ableuler I agree. I renamed and kept a small portion of this method that now only checks that the flask app config contains the necessary things. Also since both the server and user classes are so tightly coupled to jupyterhub I did not pass the config in the constructor but rather used the `current_app.config` object. However if you think that I should pass the config in the constructor to both classes I can easily make that change.

ableuler replied:
ok, it's probably the more pragmatic approach.

olevski commented on renku_notebooks/api/classes/user.py:
```diff
@@ -0,0 +1,78 @@
+
+    @property
+    def prefix(self):
+        return self.auth.api_url
```
fully removed - using from app config

olevski commented on renku_notebooks/api/auth.py:
```diff
@@ -85,47 +78,43 @@ def decorated(*args, **kwargs):
 @bp.route("oauth_callback")
 def oauth_callback():
     """Set a token in the cookie."""
+    user = User()
```
removed user class - using app config here now.

olevski commented on renku_notebooks/api/classes/server.py:
```diff
@@ -0,0 +1,399 @@
+        pods = filter_pods_by_annotations(
+            pods, {"hub.jupyter.org/servername": self.server_name}
+        )
+        if len(pods) != 1:
```
@ableuler I do not want to throw an exception here because I would only have to capture it in the blueprint (or somewhere else later) to not have a 500 error code. This way if the pod is None I know that a pod does not exist or cannot be found and I can send the proper response from the API.

olevski commented on renku_notebooks/api/auth.py:
```diff
@@ -85,47 +78,43 @@ def decorated(*args, **kwargs):
 @bp.route("oauth_callback")
 def oauth_callback():
```
@ableuler I am not sure exactly what you mean here - lets discuss briefly and I can make a separate issue for this

ableuler replied:
We still have some bits of code here and there that assume that the notebooks service has to handle oauth dances or use a JH cookie for authenticating requests. Some of this was actually still useful for quickly testing the API when develping. But now I think we should get rid these aspects. If you can create an issue for this that would be great.

olevski replied:
#591

olevski commented on renku_notebooks/api/notebooks.py:
```diff
@@ -55,83 +35,95 @@
-    servers = get_user_servers(user, namespace, project, branch, commit_sha)
-    current_app.logger.debug(servers)
-    return jsonify({"servers": servers})
+    user_pods = user.pods
```
@ableuler, I already import the `User` class in the same place where I define the `Server` class. If I then import the `Server` class where I define the `User` class so that I can have something like `user.servers` I will get a circular import. That is why I do not have an attribute of the `User` class that returns a list of server objects.

olevski replied:
What I can do is convert the pods in server classes in the `notebooks.py` script in the line immediately below your comment and then do the filtering on server objects as you suggested. This should work.

olevski replied:
@ableuler I think it is cleaner to not introduce a 2nd function to filter the user pods - one by k8s pod annotations and another by Server class attributes. And I cannot eliminate the `filter_pods_by_attributes` function because I use it in quite a few places and in some of them I do not have access to a server object. Let me know if you are ok with this.

olevski replied:
Also the k8s pod is the 'ground truth' for the information about the servers - so I think it makes sense to go straight to the source for filtering, if possible.

olevski commented on renku_notebooks/api/notebooks.py:
```diff
@@ -55,83 +35,95 @@
+        if request.args[annotation_name] is not None:
+            annotations[
+                config.RENKU_ANNOTATION_PREFIX
+                + request_annotation_xref[annotation_name]
```
@ableuler I added a schema and webargs to parse and handle the query request params. Now adding unknown parameters does not cause an attribute errors. Any unknown parameters that are passed are just ignored. And on top of that now the query params are also shown in the swagger - before they were not.

olevski commented on renku_notebooks/api/classes/server.py:
```diff
@@ -0,0 +1,399 @@
+            payload = self._get_start_payload()
+            if payload is None:
+                # a specific image was requested but does not exist
+                return make_response(
```
Agreed! Now the start method on the server class returns a tuple consisting of the JH api response (or none if things went wrong) and an error message (which is None if things worked).

ableuler reviewed (changes requested):
Closing in, I guess the biggest thing left to do is the rebasing on top of master - sorry about that 🙈.

Another small point: would be great if you could add docstrings to the new methods and functions that don't have any yet.

olevski commented on renku_notebooks/api/schemas.py:
```diff
@@ -155,6 +161,91 @@ class LaunchNotebookResponse(Schema):
+            if dt is None:
+                return None
+            dt = dt.astimezone(timezone.utc).replace(tzinfo=None)
+            return dt
```
@ableuler the marshmallow schema takes care of this when it marshals the response. If I convert to string it actually errors out because it tries to call `.isoformat()` on the python `datetime`. So I just fully removed that whole function - we do not need it.

It knows to call `isoformat` because that field is defined like this in the schema:
```
started = fields.DateTime(format="iso", allow_none=True)
```

olevski commented:
@ableuler take a look I addressed all your comments.

Rok's code I rebased and pulled into the Server class rather than leaving it in the utils. The code from Rok changed very little. The only thing that changed is that the arguments in the functions from utils now are not needed since they are part of the server class and all the arguments the functions needed are now attributes.

rokroskar commented:
Thanks for doing the rebase @olevski! 🙏

olevski commented:
I just ran the tests locally with docker and the only failing test is the one that always fails recently:
```
[info] BatchProjectRemovalSpec:
[info] Scenario: User can delete many projects project - done
[info] ScalaTest
[info] Run completed in 31 minutes, 26 seconds.
[info] Total number of tests run: 11
[info] Suites: completed 10, aborted 0
[info] Tests: succeeded 10, failed 1, canceled 0, ignored 0, pending 0
[info] *** 1 TEST FAILED ***
[error] Failed: Total 11, Failed 1, Errors 0, Passed 10
[error] Failed tests:
[error]         ch.renku.acceptancetests.MergeRequestsAndIssuesSpec
[error] (Test / test) sbt.TestsFailedException: Tests unsuccessful
[error] Total time: 1932 s (32:12), completed Mar 25, 2021, 2:32:34 PM
```

And I can never reproduce this failure when trying the same scenario by hand.

ableuler reviewed (approved):
Looks good to me! Thanks a lot for the big effort and your perseverance! 💪
